internal/net/http/transport/roundtrip.go

```diff
@@ -41,50 +42,77 @@ func NewExpBackoff(opts ...Option) http.RoundTripper {
 	return e
 }
 
+// RoundTrip round trip the HTTP request and return the response.
+// If backoff is not set, the default roundTrip implementation will be used.
+// It round trip the request and returns the response, and return any error occurred.
+// It returns errors.ErrTransportRetryable to indicate if the request is consider as retryable.
 func (e *ert) RoundTrip(req *http.Request) (res *http.Response, err error) {
 	if e.bo == nil {
 		return e.roundTrip(req)
 	}
-	var r interface{}
-	r, err = e.bo.Do(req.Context(), func() (interface{}, error) {
-		return e.roundTrip(req)
+
+	var rterr error
+	_, err = e.bo.Do(req.Context(), func() (interface{}, error) {
+		r, reqerr := e.roundTrip(req)
+		if reqerr != nil {
+			// if the error is retryable, return the error and let backoff to retry.
+			if errors.Is(reqerr, errors.ErrTransportRetryable) {
+				return nil, reqerr
+			}
+			// if the error is not retryable, return nil error to terminate the backoff execution
+			rterr = reqerr
+			return nil, nil
+		}
+		res = r
+		return r, nil
 	})
 	if err != nil {
 		return nil, err
 	}
-
-	var ok bool
-	res, ok = r.(*http.Response)
-	if !ok {
-		return nil, errors.ErrInvalidTypeConversion(r, res)
+	if rterr != nil {
+		return nil, rterr
 	}
+
 	return res, nil
 }
 
 func (e *ert) roundTrip(req *http.Request) (res *http.Response, err error) {
 	res, err = e.transport.RoundTrip(req)
-	if err == nil {
-		return res, nil
-	}
-	if res == nil {
-		return nil, err
-	}
-	_, err = io.Copy(ioutil.Discard, res.Body)
 	if err != nil {
-		log.Error(err)
+		if res != nil { // just in case we check the response as it depends on RoundTrip impl.
+			closeBody(res.Body)
+			if retryable(res) {
+				return nil, errors.Wrap(errors.ErrTransportRetryable, err.Error())
```

Comment:
Does it need to output error log before return?

Reply:
or maybe it is better to write it on line 69 and 72?

Reply:
umm, in my opinion, I prefer output between L84 and L85.

Reply:
or line 81?

Reply:
line 84 may not log the error if the response is not retryable. 🤔

Reply:
so between L81 and L82 is better?

Reply:
I think yes.

Reply:
ok

Reply:
fixed